File: apps/mobile/src/services/auth/authService.ts
import {apiClient} from '../api/apiClient';
import {User, ApiResponse} from '../../types';
import {storage} from '../../utils/storage';

interface SignInRequest {
  email: string;
  password: string;
}

interface SignUpRequest {
  email: string;
  password: string;
  name: string;
  phone?: string;
}

interface AuthResponse {
  user: User;
  accessToken: string;
  refreshToken: string;
}

/**
 * 인증 서비스
 */
export const authService = {
  /**
   * 로그인
   */
  async signIn(email: string, password: string): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/signin', {
      email,
      password,
    });

    // 토큰 저장
    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },

  /**
   * 회원가입
   */
  async signUp(data: SignUpRequest): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/signup', data);

    // 토큰 저장
    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },

  /**
   * 로그아웃
   */
  async signOut(): Promise<void> {
    await storage.clearTokens();
    // 필요시 서버에 로그아웃 요청
  },

  /**
   * 토큰 갱신
   */
  async refreshToken(): Promise<string> {
    const refreshToken = await storage.getRefreshToken();

    if (!refreshToken) {
      throw new Error('No refresh token available');
    }

    const response = await apiClient.post<ApiResponse<{accessToken: string; refreshToken: string}>>(
      '/auth/refresh',
      {refreshToken}
    );

    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data.accessToken;
  },

  /**
   * 현재 사용자 정보 조회
   */
  async getCurrentUser(): Promise<User> {
    const response = await apiClient.get<ApiResponse<User>>('/auth/me');
    return response.data;
  },

  /**
   * 소셜 로그인 - 카카오
   */
  async signInWithKakao(accessToken: string): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/kakao', {
      accessToken,
    });

    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },

  /**
   * 소셜 로그인 - 네이버
   */
  async signInWithNaver(accessToken: string): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/naver', {
      accessToken,
    });

    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },

  /**
   * 소셜 로그인 - 구글
   */
  async signInWithGoogle(idToken: string): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/google', {
      idToken,
    });

    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },

  /**
   * 소셜 로그인 - 애플
   */
  async signInWithApple(identityToken: string): Promise<AuthResponse> {
    const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/apple', {
      identityToken,
    });

    await storage.setAccessToken(response.data.accessToken);
    await storage.setRefreshToken(response.data.refreshToken);

    return response.data;
  },
};

File: apps/ios/Woorijib/Core/Network/Services/AuthService.swift
import Foundation
import Combine

protocol AuthServiceProtocol {
    var isAuthenticated: Bool { get }
    var authStatePublisher: AnyPublisher<Bool, Never> { get }

    func signIn(email: String, password: String) async throws -> User
    func signUp(email: String, password: String, name: String) async throws -> User
    func signOut()
    func getCurrentUser() async throws -> User
    func refreshToken() async throws
}

class AuthService: AuthServiceProtocol {
    static let shared = AuthService()

    private let apiClient = APIClient.shared
    private let tokenManager = TokenManager.shared
    private let authStateSubject = CurrentValueSubject<Bool, Never>(false)

    var isAuthenticated: Bool {
        tokenManager.hasValidToken
    }

    var authStatePublisher: AnyPublisher<Bool, Never> {
        authStateSubject.eraseToAnyPublisher()
    }

    private init() {
        authStateSubject.send(isAuthenticated)
    }

    func signIn(email: String, password: String) async throws -> User {
        struct SignInRequest: Encodable {
            let email: String
            let password: String
        }

        struct SignInResponse: Decodable {
            let user: User
            let accessToken: String
            let refreshToken: String

            enum CodingKeys: String, CodingKey {
                case user
                case accessToken = "access_token"
                case refreshToken = "refresh_token"
            }
        }

        let request = SignInRequest(email: email, password: password)
        let response: SignInResponse = try await apiClient.post("/auth/signin", body: request)

        tokenManager.saveTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )

        authStateSubject.send(true)

        return response.user
    }

    func signUp(email: String, password: String, name: String) async throws -> User {
        struct SignUpRequest: Encodable {
            let email: String
            let password: String
            let name: String
        }

        struct SignUpResponse: Decodable {
            let user: User
            let accessToken: String
            let refreshToken: String

            enum CodingKeys: String, CodingKey {
                case user
                case accessToken = "access_token"
                case refreshToken = "refresh_token"
            }
        }

        let request = SignUpRequest(email: email, password: password, name: name)
        let response: SignUpResponse = try await apiClient.post("/auth/signup", body: request)

        tokenManager.saveTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )

        authStateSubject.send(true)

        return response.user
    }

    func signOut() {
        tokenManager.clearTokens()
        authStateSubject.send(false)
    }

    func getCurrentUser() async throws -> User {
        try await apiClient.get("/auth/me")
    }

    func refreshToken() async throws {
        guard let refreshToken = tokenManager.refreshToken else {
            throw AuthError.noRefreshToken
        }

        struct RefreshRequest: Encodable {
            let refreshToken: String

            enum CodingKeys: String, CodingKey {
                case refreshToken = "refresh_token"
            }
        }

        struct RefreshResponse: Decodable {
            let accessToken: String
            let refreshToken: String

            enum CodingKeys: String, CodingKey {
                case accessToken = "access_token"
                case refreshToken = "refresh_token"
            }
        }

        let request = RefreshRequest(refreshToken: refreshToken)
        let response: RefreshResponse = try await apiClient.post("/auth/refresh", body: request)

        tokenManager.saveTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )
    }
}

enum AuthError: LocalizedError {
    case noRefreshToken
    case invalidCredentials

    var errorDescription: String? {
        switch self {
        case .noRefreshToken:
            return "인증 토큰이 없습니다."
        case .invalidCredentials:
            return "이메일 또는 비밀번호가 올바르지 않습니다."
        }
    }
}
